core/src/main/java/org/mule/runtime/core/execution/ModuleFlowProcessingPhase.java

```diff
@@ -219,11 +221,24 @@ public void runPhase(final ModuleFlowProcessingPhaseTemplate template, final Mes
     }
   }
 
-  private void handleSourceError(final MessagingExceptionHandler exceptionHandler, Consumer<MessagingException> errorConsumer,
-                                 SourceErrorException see) {
+  private Mono<? extends Void> onSourceException(MessagingExceptionHandler exceptionHandler,
+                                                 Function<MessagingException, Publisher<Void>> errorHandler,
+                                                 Consumer<Either<Event, MessagingException>> terminateConsumer,
+                                                 SourceErrorException see) {
+    if (sourceResponseErrorTypeMatcher.match(see.getErrorType())) {
+      return from(handleSourceError(exceptionHandler, errorHandler, see))
+          .doOnSuccess(v -> onTerminate(terminateConsumer, left(see.getEvent())))
```

Comment:
`onTerminate` should be being invoked from main Mono, not in onSucess/onError independently.